src/Orleans.Runtime/Streams/PersistentStream/PersistentStreamPullingAgent.cs

```diff
@@ -606,13 +606,28 @@ private async Task RunConsumerCursor(StreamConsumerData consumerData, IStreamFil
 
         private async Task<StreamHandshakeToken> DeliverBatchToConsumer(StreamConsumerData consumerData, IBatchContainer batch)
         {
-            StreamHandshakeToken prevToken = consumerData.LastToken;
-            Task<StreamHandshakeToken> batchDeliveryTask;
+            try
+            {
+                StreamHandshakeToken newToken = await ContextualizedDeliverBatchToConsumer(consumerData, batch);
+                consumerData.LastToken = StreamHandshakeToken.CreateDeliveyToken(batch.SequenceToken); // this is the currently delivered token
+                return newToken;
+            }
+            catch (Exception ex)
+            {
+                this.logger.LogWarning(ex, "Failed to deliver message to consumer on {SubscriptionId} for stream {StreamId}, may retry.", consumerData.SubscriptionId, consumerData.StreamId);
```

Comment:
Logging on each attempted delivery failure within a retry loop concerns me.  In local tests this logs 10-20 times before retry logic gives up.  Since this logging is per consumer, I worry this could spam logs.

Preexisting logic only logged once retries were exhausted, at the cost of only being able to log the exception of the last failed call.

Reply:
valid concern. I guess we can have one escape plan when it proven to be spamming. One way is give this logger a special logger name. And if it is too spamming, application code can filter it out.

Reply:
or , we can give this logging message a specific event ID. So if the application find this messaging too spamming, they can filter it out in their logger implementation. Unfortunately, filter out a certain event ID isn't included in our logging abstraction. So logger implementation out there doesn't take this into account. If application want to filter out certain event ID, then it need some custom logic in their logger.

Reply:
@jason-bragg I'm not sure it would be spam... It is a real problem and must be logged... People using structured logs can even count and have metrics based on that... I would write everything...